[WIP] CMF / Dynamic Routing

aschempp opened the pull request:
> Continuing from contao/core-bundle#1653

This is a work in progress how to implement the CMF Dynamic Router to the Contao code base. It currently does nothing more than routing, there are no new features like real controllers for pages etc.

Feel free to review and test it, but there's a lot still to be changed 😊

**TODO:**
- [ ] If `prepend_locale` is enabled and a page without locale is accessed, the visitor was redirected. E.g. https://demo.contao.org/news.html => https://demo.contao.org/en/news.html. 
- [x] If no pages are defined in `tl_page`, the `NoRootPageFoundException` should be rendered
- [x] Because the `PrettyErrorScreenHandler` only handles Contao routes (`_scope=frontend/backend`), it is now never triggered on actual 404s.
- [x] ~should we add a redirect if root is set to use HTTPS?~ see #262 
- [x] is `/` really a valid alternative alias for the first page instead of `index`?

aschempp committed:
Rebase from contao/core-bundle

aschempp committed:
Fixed url suffix in FrontendLoader test

aschempp committed:
Only initialize the route-related Contao framework after routing

aschempp committed:
Fixed other unit tests

aschempp committed:
Do not force SSL (yet) if enabled in the root page

aschempp committed:
Correctly test the request instead of the request stack in framework

aschempp committed:
Fixed auto_item handling

aschempp committed:
Fixed alias matching if folderUrls is disabled with getPageIdFromUrl hook

aschempp committed:
Drop useless outdated test

aschempp committed:
Load the CmfRoutingBundle in the functional tests

aschempp committed:
Correctly decode the request

aschempp commented:
To "successfully" run the functional tests, the following adjustments to core are necessary:
 - Remove `exit` in `Controller::redirect()`
 - Replace the `routing.yml` in functional tests with the one from `Resources/config`

The tests are successful so far as the following things are *up for discussion*:
 1. `PageRoot::generate()` throws a 303 instead of 302 redirect. This code was probably not reached previously because the URL matching already redirected, but Sf routing now simply uses the correct controller.
 2. The alias `/.html` or `/en/.html` was previously redirected to the first page. I think that's wrong, it should generate a 404 (Sf routing does that).
 3. The "no root page found" state currently renders a 404 instead of a 500. And probably does not render the correct "information page", because Sf just does not find a matching route. Not sure how we should handle that, because we would need to have a controller for that case.

leofeyer commented on core-bundle/src/Framework/ContaoFramework.php:
```diff
@@ -80,9 +75,8 @@ class ContaoFramework implements ContaoFrameworkInterface, ContainerAwareInterfa
     private $hookListeners = [];
 
-    public function __construct(RequestStack $requestStack, RouterInterface $router, ScopeMatcher $scopeMatcher, string $rootDir, int $errorLevel)
+    public function __construct(?RequestStack $requestStack /* removed in Contao 4.7 */, RouterInterface $router, ScopeMatcher $scopeMatcher, string $rootDir, int $errorLevel)
```
Why did you comment this instead of removing it?

aschempp replied:
Because it would be a BC break?

leofeyer replied:
Hm, the parameter has been removed in Contao 4.7 and your pull request is targeted against Contao 4.7. 🤔

aschempp replied:
What do you mean it has been removed?

leofeyer replied:
That's what your comment says: `/* removed in Contao 4.7 */`

aschempp replied:
lol. Yeah, it was removed **in this pull request** because we now have `setRequest`. But we can't change the arguments, that's why there is a comment on why the argument is still there.

leofeyer commented on core-bundle/src/Framework/ContaoFramework.php:
```diff
@@ -127,6 +118,25 @@ public function setHookListeners(array $hookListeners): void
         $this->hookListeners = $hookListeners;
     }
 
+    public function setRequest(Request $request)
```
Why is this necessary?

aschempp replied:
If the framework is bootest before routing (a listener with a higher priority) we cannot initialize some parts of the framework.

leofeyer commented on core-bundle/src/Framework/ContaoFramework.php:
```diff
@@ -158,26 +168,36 @@ public function getAdapter($class): Adapter
+            \define('TL_START', microtime(true));
+        }
+
+        if (!\defined('TL_ROOT')) {
```
This check should not have to be there, as the framework cannot be initialized twice. Did you add it for the unit tests only?

aschempp replied:
`TL_ROOT` cannot be initialized if the request is not matched.

leofeyer commented on core-bundle/src/Routing/FrontendLoader.php:
```diff
@@ -61,7 +67,7 @@ public function supports($resource, $type = null): bool
     private function addFrontendRoute(RouteCollection $routes, array $defaults): void
     {
-        $route = new Route('/{alias}%contao.url_suffix%', $defaults, ['alias' => '.+']);
+        $route = new Route('/{alias}'.$this->urlSuffix, $defaults, ['alias' => '.+']);
```
What was the problem with the previous implementation?

aschempp replied:
Because we're no longer using the route loader but manually use the class (to keep BC) the container parameters are not resolved. see https://github.com/contao/contao/pull/95/files#diff-5f8c741675801f711d0a61e756575413R51

leofeyer commented on core-bundle/src/Routing/LegacyRouteProvider.php:
```diff
@@ -0,0 +1,89 @@
+     */
+    public function getRouteByName($name)
+    {
+        if ('contao_frontend' === $name || 'contao_index' === $name) {
```
Why can we no longer have these routes in the `routing.yml` file?

aschempp replied:
They must not be matched by the routing, but they must be available to the URL generator.

leofeyer reviewed (changes requested):
It seems to me that you have made some changes for the unit tests only, which is the wrong approach IMHO.

aschempp committed:
Only initialize constants if request is available

leofeyer commented:
> If `prepend_locale` is enabled and a page without locale is accessed, the visitor was redirected. E.g. https://demo.contao.org/news.html => https://demo.contao.org/en/news.html. Is that intended?

It sure is. I have added a test in https://github.com/contao/contao/commit/825a5b095b84415b88645a38dc3bbcfe8e5a9d85.

ausi commented:
This redirect is currently also triggered for non-existent pages (e.g. <https://demo.contao.org/foo.html> redirects to <https://demo.contao.org/en/foo.html>). IMO it would be better to show a 404 page directly instead of the redirect.

leofeyer commented:
> To "successfully" run the functional tests, the following adjustments to core are necessary:
> - Remove `exit` in `Controller::redirect()`
> - Replace the `routing.yml` in functional tests with the one from `Resources/config`

Done.

> The tests are successful so far as the following things are *up for discussion*:
> 1. `PageRoot::generate()` throws a 303 instead of 302 redirect. This code was probably not reached previously because the URL matching already redirected, but Sf routing now simply uses the correct controller.

Fine with me.

>  2. The alias `/.html` or `/en/.html` was previously redirected to the first page. I think that's wrong, it should generate a 404 (Sf routing does that).

Perfectly fine!

>  3. The "no root page found" state currently renders a 404 instead of a 500. And probably does not render the correct "information page", because Sf just does not find a matching route. Not sure how we should handle that, because we would need to have a controller for that case.

As discussed in Mumble on October 18th, it would be good to preserve the "no root page found" error if possible, so beginners know why they don't see anything.

aschempp committed:
Correctly sort pages by accepted languages

aschempp committed:
Send 302 status code when redirecting a root page

aschempp commented:
PR updated to correctly sort/match pages based on the browser accept language. That was not present yet 😇

aschempp commented:
> As discussed in Mumble on October 18th, it would be good to preserve the "no root page found" error if possible, so beginners know why they don't see anything.

This is possible, however, the `PrettyErrorScreenListener` does not render at all if the pages are not in `scope = frontend/backend` (https://github.com/contao/contao/blob/master/core-bundle/src/EventListener/PrettyErrorScreenListener.php#L106). And if the router does not match a page, there is no scope, so it never renders. Can we agree on changing this so the pretty screens are always rendered for HTML output? @sheeep @bytehead

leofeyer commented:
I'm not sure about it, but if it is the only solution, I'd rather have it this way than not having it at all.

bytehead commented:
The `PrettyErrorScreenListener` is fine for me for routes in Contao scope, but only for them. 

> Can we agree on changing this so the pretty screens are always rendered for HTML output?

👎

aschempp commented:
> The `PrettyErrorScreenListener` is fine for me for routes in Contao scope, but only for them.

How do you want to know if a 404 (no match found) is Contao route?

leofeyer commented:
Yeah, that seems impossible. You can either enable pretty error screens – then you will get them everywhere – or you can disable them – then you will not get them anywhere.

leofeyer commented:
@aschempp Regarding your new questions:

> should we add a redirect if root is set to use HTTPS?

Yes, that would be nice. 👍 

> is / really a valid alternative alias for the first page instead of index?

I guess so, as some servers distinct between `/home` and `/home/` (aka with or without trailing slash).

aschempp commented:
> should we add a redirect if root is set to use HTTPS?

I think that should be a new feature that we implement separately from this PR

> > is / really a valid alternative alias for the first page instead of index?

> I guess so, as some servers distinct between /home and /home/ (aka with or without trailing slash).

I don't think you got this right. The alias "/" (no other content) is matched as the index page in the Contao 3 router.

leofeyer commented:
Yes, I know. This is intentionally.

aschempp commented:
I guess this is *up for discussion* again as we need to agree about the `PrettyErrorScreenListener`

bytehead commented:
> Yeah, that seems impossible. You can either enable pretty error screens – then you will get them everywhere – or you can disable them – then you will not get them anywhere.

If the `PrettyErrorScreenListener` is disabled by default, I'm fine with this then.

aschempp commented:
> If the PrettyErrorScreenListener is disabled by default, I'm fine with this then.

How can it be disabled by default? A Contao 404 page would never be rendered, which is a basic feature of every Contao website.

bytehead commented:
> How can it be disabled by default?

It has to be disabled while using Contao just as a Symfony bundle.
I don't care what's configured by default in the _Managed Edition_.

> A Contao 404 page would never be rendered, which is a basic feature of every Contao website.

What if a 404 page (type) is configured in the page tree?

aschempp commented:
> What if a 404 page (type) is configured in the page tree?

Thats what I'm talking about. It would never be rendered …

bytehead commented:
Why not? If you have a configured 404 page in the page tree, you don't need to have the `PrettyErrorScreenListener`, don't you?

aschempp commented:
Yes you do. The `PrettyErrorScreenHandler` catches `ResourceNotFoundException` or `HttpNotFoundException` and renders the page from page tree if available 😉

bytehead commented:
Ok. Fine for me in the _Managed Edition_. I don't like to leak Contao too much into Symfony as just a bundle (as we use it in all Contao cases).

If I'm fine with the Contao 404, I'll have to configure `prettyErrorScreen: true`, by default I get the default Symfony one.

leofeyer commented:
As discussed in Mumble on November, 22rd:

> [15:44] (Channel) andy: es ist ganz einfach
> [15:44] (Channel) andy: core-bundle default aus
> [15:44] (Channel) andy: manager-bundle schaltet es ein

aschempp committed:
Render pretty error screens only in Contao Managed Edition

leofeyer commented:
@aschempp Can you please fix the existing unit tests and add the missing ones (see coverage)? Then I can merge this PR.

aschempp committed:
Remove test for TL_START

aschempp committed:
Test RootPageNotFound in the exception converter listener

aschempp committed:
Test InputEnhancer and added some TODOs

aschempp committed:
InputEnhancer must also decode the $_GET value

aschempp committed:
Added unit tests for PublishingFilter and LegacyMatcher

aschempp committed:
Fixed merge conflicts in ContaoFrameworkTest

aschempp committed:
Added initial tests for RouteProvider

aschempp committed:
Fixed unit tests in ContaoFramework

leofeyer commented on core-bundle/src/Resources/config/listener.yml:
```diff
@@ -70,10 +70,21 @@ services:
     contao.listener.exception_converter:
         class: Contao\CoreBundle\EventListener\ExceptionConverterListener
+        arguments:
+            - "@database_connection"
```
I assume you have check that we are not ending in a loop if the database class throws an exception while we are converting the exception?

aschempp replied:
I have not, but I'm pretty sure the Symfony exception handler handles this case. But feel free to verify.

leofeyer commented on core-bundle/src/Resources/contao/classes/Frontend.php:
```diff
@@ -59,6 +59,8 @@ public function __construct()
 	 */
```
The phpDoc is missing a `@deprecated` tag (this is just a note to myself).

aschempp replied:
Added in 847914e2ed2909575f53e21e66a8e9f42ed40004

leofeyer commented on core-bundle/src/Routing/Enhancer/InputEnhancer.php:
```diff
@@ -0,0 +1,99 @@
+            return $defaults;
+        }
+
+        // TODO should we use trim() instead of substr()?
```
Yes, `ltrim()` I guess.

aschempp replied:
As discussed, this should not be changed to not remove multiple slashes on the beginning of the path.

leofeyer commented on core-bundle/src/Routing/Enhancer/InputEnhancer.php:
```diff
@@ -0,0 +1,99 @@
+
+        for ($i = 0, $c = \count($fragments); $i < $c; $i += 2) {
+            // Skip key value pairs if the key is empty (see #4702)
+            // TODO why does this make sense?
```
I guess that is the double slash case `news//something.html`. I will check if the functional tests handle this case.

aschempp replied:
see #263

leofeyer commented on core-bundle/src/Routing/Enhancer/InputEnhancer.php:
```diff
@@ -0,0 +1,99 @@
+            }
+
+            // Abort if there is a duplicate parameter (duplicate content) (see #4277)
+            // TODO should we use $request->query here?
```
Yes

aschempp replied:
See 847914e2ed2909575f53e21e66a8e9f42ed40004

leofeyer commented on core-bundle/src/Routing/Matcher/LegacyMatcher.php:
```diff
@@ -0,0 +1,196 @@
+    private function executeLegacyHook(array $fragments)
+    {
+        foreach ($GLOBALS['TL_HOOKS']['getPageIdFromUrl'] as $callback) {
+            /** @noinspection StaticInvocationViaThisInspection */
```
I think we have agreed on not adding any IDE specific code. 🙈

aschempp replied:
Removed in 847914e2ed2909575f53e21e66a8e9f42ed40004

leofeyer commented on core-bundle/src/Routing/RouteProvider.php:
```diff
@@ -0,0 +1,472 @@
+
+            [, $id] = explode('.', $name);
+
+            $ids[] = $id;
```
We should type-cast to `(int)` here.

aschempp replied:
Implemented in 847914e2ed2909575f53e21e66a8e9f42ed40004

leofeyer commented on core-bundle/src/Routing/RouteProvider.php:
```diff
@@ -0,0 +1,472 @@
+            $requirements,
+            ['utf8' => true],
+            $page->domain,
+            null // TODO should we match SSL only if enabled in root? => $page->rootUseSSL ? 'https' : null
```
I would prefer to automatically redirect HTTP to HTTPS.

aschempp replied:
see #262

leofeyer commented on core-bundle/tests/Framework/ContaoFrameworkTest.php:
```diff
@@ -277,6 +260,10 @@ public function testOverridesTheErrorLevel(): void
     }
 
+    /**
+     * @runInSeparateProcess
```
Why did you re-add this? The tests in the master branch work without a separate process.

aschempp committed:
Resolve TODOs and fix issues found by leofeyer

https://github.com/contao/contao/issues/262
https://github.com/contao/contao/issues/263
https://github.com/contao/contao/issues/264

aschempp committed:
Test that the FrontendLoader adds the url suffix

asaage commented:
Can this accomplish correct error-pages for sites with enabled folder-url's?
https://github.com/contao/core-bundle/issues/1624

leofeyer committed:
Fix the coding style

leofeyer committed:
Merge branch 'master' into feature/dynamic-router

aschempp committed:
Correctly initialize and validate the locale string

aschempp committed:
Fixed routing functional tests

leofeyer commented on core-bundle/tests/Functional/RoutingTest.php:
```diff
@@ -189,8 +189,8 @@ public function getAliases(): array
             ],
             'Renders the default page if the alias is empty' => [
```
Renders the 404 page if the alias is empty

aschempp replied:
Fixed in 5496681b13cec9843806184597d48623796eab1c

leofeyer commented on core-bundle/tests/Functional/RoutingTest.php:
```diff
@@ -412,9 +412,9 @@ public function getAliasesWithLocale(): array
             ],
             'Renders the default page if the alias is empty' => [
```
Renders the 404 page if the alias is empty

aschempp replied:
Fixed in 5496681b13cec9843806184597d48623796eab1c

leofeyer commented on core-bundle/tests/Functional/RoutingTest.php:
```diff
@@ -770,8 +770,8 @@ public function getRootAliases(): array
             ],
             'Throws a "no root page found" exception if no language matches' => [
```
Renders the 404 page…

aschempp replied:
Fixed in 5496681b13cec9843806184597d48623796eab1c

aschempp committed:
Remove unused scope matcher in PrettyErrorScreenListener

aschempp committed:
Trigger warning that request stack should not be set in ContaoFramework

aschempp committed:
Fixed names of dataProvider test cases

aschempp committed:
Rewrite Frontend::getRootPageFromUrl to use the request matcher

leofeyer commented on core-bundle/src/Resources/contao/classes/Frontend.php:
```diff
@@ -291,84 +299,71 @@ public static function getRootIdFromUrl()
 	public static function getRootPageFromUrl()
 	{
-		// HOOK: add custom logic
-		if (isset($GLOBALS['TL_HOOKS']['getRootPageFromUrl']) && \is_array($GLOBALS['TL_HOOKS']['getRootPageFromUrl']))
```
Removing the hook seems like a BC break to me.

aschempp replied:
The hook is triggered by the the new route provider (https://github.com/contao/contao/pull/95/files#diff-8f3e1abe0064e8534f4e9193ef359276R470)

leofeyer committed:
Fix the coding style and the unit tests

aschempp committed:
Fixed unit tests and expected deprecations

leofeyer committed:
Fix the functional tests

aschempp committed:
Automatically import routing SQL dump and test for no root page

aschempp committed:
Renamed mockFramework method because it does not create a mock

aschempp committed:
Do not use a mock to test a class

aschempp committed:
Added test that the framework is initialized after the request is set

aschempp committed:
Updated the composer.json

leofeyer committed:
Prepare the merge

leofeyer commented:
Thanks a lot @aschempp. This is huge! 🎉
